test/lit/passes/simplify-locals-strings.wast

```diff
@@ -0,0 +1,544 @@
+  ;; CHECK-NEXT:   (local.get $array)
+  ;; CHECK-NEXT:   (i32.const 3)
+  ;; CHECK-NEXT:   (i32.const 4)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $no-new-past-store-gc (param $array (ref $array)) (param $array16 (ref $array16))
+    (local $temp stringref)
+    ;; A string.new cannot be moved past a GC store.
+    (local.set $temp
+      (string.new_wtf8_array utf8
+        (local.get $array)
+        (i32.const 1)
+        (i32.const 2)
+      )
+    )
+    (array.set $array
+      (local.get $array)
+      (i32.const 3)
+      (i32.const 4)
+    )
+    (drop
+      (local.get $temp)
+    )
+    (local.set $temp
+      (string.new_wtf8_array wtf8
+        (local.get $array)
+        (i32.const 1)
+        (i32.const 2)
+      )
+    )
+    (array.set $array
+      (local.get $array)
+      (i32.const 3)
+      (i32.const 4)
+    )
+    (drop
+      (local.get $temp)
+    )
+    (local.set $temp
+      (string.new_wtf8_array replace
+        (local.get $array)
+        (i32.const 1)
+        (i32.const 2)
+      )
+    )
+    (array.set $array
+      (local.get $array)
+      (i32.const 3)
+      (i32.const 4)
+    )
+    (drop
+      (local.get $temp)
+    )
+    (local.set $temp
+      (string.new_wtf16_array
+        (local.get $array16)
+        (i32.const 1)
+        (i32.const 2)
+      )
+    )
+    (array.set $array
+      (local.get $array)
+      (i32.const 3)
+      (i32.const 4)
+    )
+    (drop
+      (local.get $temp)
+    )
+  )
+
+  ;; CHECK:      (func $no-load-past-encode (param $ref stringref)
+  ;; CHECK-NEXT:  (local $temp i32)
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (string.encode_wtf8 wtf8
+  ;; CHECK-NEXT:    (local.get $ref)
+  ;; CHECK-NEXT:    (i32.const 10)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (string.encode_wtf8 utf8
+  ;; CHECK-NEXT:    (local.get $ref)
+  ;; CHECK-NEXT:    (i32.const 20)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.set $temp
+  ;; CHECK-NEXT:   (i32.load
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (string.encode_wtf16
+  ;; CHECK-NEXT:    (local.get $ref)
+  ;; CHECK-NEXT:    (i32.const 30)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (local.get $temp)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $no-load-past-encode (param $ref stringref)
```

Comment:
Why can't this go past the encode? Does `encode` operation only read memory or write to memory as well? It'd be helpful to have one-line comments in this function and the functions below as we did for previous functions.

Reply:
Encode takes an existing String and writes it out into either linear memory or an array. So it has write effects.

Added comments now.